### PR TITLE
Add unary_cmp

### LINKED ISSUE #1987: Kernel to quickly compute comparisons of arrays
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
DataFusion (and likely other libraries) often wants to implement its own fast comparison functions 

For example, comparing the contents of an array element by element to form an output `BooleanArray`
https://github.com/apache/arrow-datafusion/blob/7617d78809d4ff5bde31142e0744c70024e40635/datafusion/physical-expr/src/expressions/in_list.rs#L50

the `arity::unary` is close to what I want: https://docs.rs/arrow/17.0.0/arrow/compute/kernels/arity/index.html but it can not produce a `BooleanArray` (because` BooleanArray` is not a `PrimitiveArray`)

```rust
fn in_list_primitive<T: ArrowPrimitiveType>(
    array: &PrimitiveArray<T>,
    values: &[<T as ArrowPrimitiveType>::Native],
) -> Result<BooleanArray> {
    Ok(arity::unary(array, |x| values.contains(&x)))
}
```

**Describe the solution you'd like**
I would like a function such as `unary_cmp` that I could call like:

```rust

fn in_list_primitive<T: ArrowPrimitiveType>(
    array: &PrimitiveArray<T>,
    values: &[<T as ArrowPrimitiveType>::Native],
) -> Result<BooleanArray> {
    Ok(arity::unary_cmp(array, |x| values.contains(&x)))
}

```

There is a proposed implementation: https://github.com/apache/arrow-datafusion/pull/2826

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1987.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
Unrelated error:

```
/usr/bin/ld: final link failed: No space left on device
```

**Review by tustvold — APPROVED** [review]
Perhaps we could inline the macro in unary_op and make the other functions use it 🤔. Less macros is always good imo

**Comment by viirya** [comment]
> Perhaps we could inline the macro in unary_op and make the other functions use it 🤔. Less macros is always good imo

Good idea!

File: arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -2554,7 +2554,16 @@ where
     #[cfg(feature = "simd")]
     return simd_compare_op_scalar(left, right, T::eq, |a, b| a == b);
     #[cfg(not(feature = "simd"))]
-    return compare_op_scalar!(left, right, |a, b| a == b);
+    return compare_op_scalar!(left, |a| a == right);
+}
+
+/// Applies an unary and infallible comparison function to a primitive array.
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:2560]: ❤️  thank you @viirya

**Pull request merged**